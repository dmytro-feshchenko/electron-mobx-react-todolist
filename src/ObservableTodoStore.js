import shortid from 'shortid';
import mobx, { observable, computed } from 'mobx';

/**
 * ObservableTodoStore - provides core methods to interact with
 * tasks (such as creating, editing, removing, etc.)
 *
 * Example:
 *      observableTodoStore.addTodo("Buy a bottle of milk in Wallmart after work")
 */
class ObservableTodoStore {
    @observable todos = [];
    @observable pendingRequests = 0;
    
    constructor() {
        // shows report about current tasks in console after this.todos changing
        // automatically
        mobx.autorun(() => console.log(this.report));
    }

    /**
     * Returns count of all completed tasks
     * @returns {Number}
     */
    @computed get completedTodosCount() {
        return this.todos.filter(
            todo => todo.completed === true
        ).length;
    }

    /**
     * Returns common report about current tasks
     * If there are some tasks - shows statistics and next task.
     * Otherwise, shows a message with instructions "How to Start".
     * @returns {string}
     */
    @computed get report() {
        if (this.todos.length === 0)
            return "No tasks yet. You can create a new task using the button below.";

        return `Next todo: "${this.todos[0].task}". ` +
            `Progress: ${this.completedTodosCount}/${this.todos.length}`;
    }

    /**
     * Adds a new task to the list
     * @param text  Message of the task
     */
    addTodo(text) {
        this.todos.push({
            id: shortid.generate(),
            text: text,
            completed: false,
            assignee: null
        })
    }
}

export default ObservableTodoStore;