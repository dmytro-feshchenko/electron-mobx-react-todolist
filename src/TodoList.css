.TodoList-add-button {
    background: #222;
    color: white;
    font-size: 14px;
    outline: none;
    border: none;
    padding: 0 15px;
    cursor: pointer;
    line-height: 34px;
    vertical-align: bottom;
}

.TodoList-create-task-form {
    margin: 10px auto;
}

.TodoList-create-task-text-field {
    border: 1px solid #222;
    margin-right: 5px;
    line-height: 32px;
    min-width: 200px;
    padding: 0 10px;
}