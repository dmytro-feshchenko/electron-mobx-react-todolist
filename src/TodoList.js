import React, { Component } from 'react';
import TodoView from './TodoView';
import './TodoList.css';

/**
 * TodoList - represents an interactive list
 * which shows all todos created by a user
 */
class TodoList extends Component {

    render() {
        const store = this.props.store;
        return (
            <div>
                {store.report}
                <ul>
                    { store.todos.map(
                        (todo, idx) => <TodoView todo={todo} key={todo.id} />
                    ) }
                </ul>
                <button onClick={ this.onNewTodo }>New Todo</button>
            </div>
        )
    }

    onNewTodo = () => {
        this.props.store.addTodo(prompt('Enter a new todo:','coffee plz'));
    }
}

export default TodoList;