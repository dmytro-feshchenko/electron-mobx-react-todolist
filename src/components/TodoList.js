import React, { Component } from 'react';
import TodoView from './TodoView';
import './TodoList.css';

/**
 * TodoList - represents an interactive list
 * which shows all todos created by a user
 */
class TodoList extends Component {
    /**
     * Constructor for TodoList component
     * @param props
     */
    constructor(props) {
        super(props);

        this._onNewTodo = this._onNewTodo.bind(this);
        this._onChangeNewTodoText = this._onChangeNewTodoText.bind(this);

        this.state = {
            todoText: ""
        }
    }

    /**
     * Renders the TodoList
     * @returns {XML}
     */
    render() {
        const store = this.props.store;
        return (
            <div>
                <div className="TodoList-create-task-form">
                    <input type="text"
                           className="TodoList-create-task-text-field"
                           placeholder="Type Task Name Here..."
                           onChange={this._onChangeNewTodoText} />
                    <button className="TodoList-add-button" onClick={ this._onNewTodo }>Create</button>
                </div>
                <ul>
                    { store.todos.map(
                        (todo, idx) => <TodoView todo={todo} key={todo.id} />
                    ) }
                </ul>
                {store.report}
            </div>
        )
    }

    _onChangeNewTodoText(event) {
        this.setState({todoText: event.target.value});
    }

    _onNewTodo() {
        if (this.state.todoText !== "") {
            this.props.store.addTodo(this.state.todoText);
        }
    }
}

export default TodoList;