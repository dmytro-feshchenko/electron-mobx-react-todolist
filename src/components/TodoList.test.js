import React from 'react';
import TodoList from './TodoList';
import ObservableTodoStore from '../store/ObservableTodoStore';
import renderer from 'react-test-renderer';

it("renders todos correctly", () => {
    let store = new ObservableTodoStore();
    store.addTodo("Reminder to buy milk after work");
    store.addTodo("Please, finish the report until tomorrow");
    store.addTodo("Complete online shopping transaction");

    store.todos[0].completed = true;

    const tree = renderer.create(
        <TodoList store={store} />
    );

    expect(tree).toMatchSnapshot();
});