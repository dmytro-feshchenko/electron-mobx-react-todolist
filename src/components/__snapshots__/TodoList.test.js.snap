// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders todos correctly 1`] = `
<div>
  <div
    className="TodoList-create-task-form"
  >
    <input
      className="TodoList-create-task-text-field"
      onChange={[Function]}
      placeholder="Type Task Name Here..."
      type="text"
    />
    <button
      className="TodoList-add-button"
      onClick={[Function]}
    >
      Create
    </button>
  </div>
  <ul>
    <li
      onDoubleClick={[Function]}
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      Reminder to buy milk after work
    </li>
    <li
      onDoubleClick={[Function]}
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      Please, finish the report until tomorrow
    </li>
    <li
      onDoubleClick={[Function]}
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      Complete online shopping transaction
    </li>
  </ul>
  Next todo: "Reminder to buy milk after work". Progress: 1/3
</div>
`;
